src/components/tx/SignOrExecuteForm/hooks.ts

```diff
@@ -0,0 +1,87 @@
+import { useMemo } from 'react'
+import { type TransactionOptions, type SafeTransaction } from '@safe-global/safe-core-sdk-types'
+import useSafeInfo from '@/hooks/useSafeInfo'
+import useWallet from '@/hooks/wallets/useWallet'
+import useOnboard from '@/hooks/wallets/useOnboard'
+import { isSmartContractWallet } from '@/hooks/wallets/wallets'
+import {
+  dispatchOnChainSigning,
+  dispatchTxExecution,
+  dispatchTxProposal,
+  dispatchTxSigning,
+} from '@/services/tx/tx-sender'
+import { useHasPendingTxs } from '@/hooks/usePendingTxs'
+import { sameString } from '@safe-global/safe-core-sdk/dist/src/utils'
+
+export const useTxActions = () => {
+  const { safe, safeAddress } = useSafeInfo()
+  const { chainId, version } = safe
+  const onboard = useOnboard()
+  const wallet = useWallet()
+
+  return useMemo(() => {
+    const proposeTx = async (sender: string, safeTx: SafeTransaction, txId?: string, origin?: string) => {
+      const tx = await dispatchTxProposal({
+        chainId,
+        safeAddress,
+        sender,
+        safeTx,
+        txId,
+        origin,
+      })
+      return tx.txId
+    }
+
+    return {
+      async signTx(safeTx: SafeTransaction, txId?: string): Promise<string> {
+        if (!onboard) throw 'Onboard not ready'
+        if (!wallet) throw 'Wallet not connected'
+
+        // Smart contract wallets must sign via an on-chain tx
+        if (await isSmartContractWallet(wallet)) {
+          // If the first signature is a smart contract wallet, we have to propose w/o signatures
+          // Otherwise the backend won't pick up the tx
+          // The signature will be added once the on-chain signature is indexed
+          const id = txId || (await proposeTx(wallet.address, safeTx, txId, origin))
+          await dispatchOnChainSigning(safeTx, id, onboard, chainId)
+          return id
+        }
+
+        // Otherwise, sign off-chain
+        const signedTx = await dispatchTxSigning(safeTx, version, onboard, chainId, txId)
+        return await proposeTx(wallet.address, signedTx, txId, origin)
+      },
+
+      async executeTx(
+        safeTx: SafeTransaction,
+        txId: string | undefined,
+        txOptions: TransactionOptions,
+      ): Promise<string> {
+        if (!onboard) throw 'Onboard not ready'
+        if (!wallet) throw 'Wallet not connected'
+
+        const id = txId || (await proposeTx(wallet.address, safeTx, txId, origin))
+        await dispatchTxExecution(safeTx, txOptions, id, onboard, chainId)
+        return id
+      },
+    }
+  }, [chainId, safeAddress, version, onboard, wallet])
+}
+
+export const useValidateNonce = (safeTx?: SafeTransaction): boolean => {
+  const { safe } = useSafeInfo()
+  return !!safe && !!safeTx && safeTx?.data.nonce === safe.nonce
```

Comment:
Similarly here we don't need to check for the existance of `safe` and could keep the shorthand syntax from before since the default fallback nonce is -1 which no tx should ever have.